cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,710 @@
+    bool sorted =
+      thrust::is_sorted(handle.get_thrust_policy(),
+                        matrix_partition.get_offsets(),
+                        matrix_partition.get_offsets() + matrix_partition.get_major_size() + 1);
+    auto segment_offsets = graph_view.get_local_adj_matrix_partition_segment_offsets(i);
+    auto use_dcs         = segment_offsets
+                             ? ((*segment_offsets).size() >
+                        (cugraph::detail::num_sparse_segments_per_vertex_partition + 1))
+                             : false;
+
+    point(handle, "is_sorted ", static_cast<int>(sorted), i, use_dcs);
+  }
+}
+}  // namespace debug
+#endif
+namespace cugraph {
+
+namespace detail {
+
+/**
+ * @brief Calculate local out degrees of the sources belonging to the adjacency matrices
+ * stored on each gpu
+ *
+ * Iterate through partitions and store their local degrees
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @return A single vector containing the local out degrees of the sources belong to the adjacency
+ * matrices
+ */
+template <typename GraphViewType>
+rmm::device_uvector<typename GraphViewType::edge_type> calculate_local_degrees(
+  raft::handle_t const& handle, GraphViewType const& graph_view)
+{
+  static_assert(GraphViewType::is_adj_matrix_transposed == false);
+  using vertex_t = typename GraphViewType::vertex_type;
+  using edge_t   = typename GraphViewType::edge_type;
+  using weight_t = typename GraphViewType::weight_type;
+
+  rmm::device_uvector<edge_t> local_degrees(
+    graph_view.get_number_of_local_adj_matrix_partition_rows(), handle.get_stream());
+
+  // get_number_of_local_adj_matrix_partition_rows == summation of get_major_size() of all
+  // partitions belonging to the gpu
+  vertex_t partial_offset{0};
+  for (size_t i = 0; i < graph_view.get_number_of_local_adj_matrix_partitions(); ++i) {
+    auto matrix_partition =
+      matrix_partition_device_view_t<vertex_t, edge_t, weight_t, GraphViewType::is_multi_gpu>(
+        graph_view.get_matrix_partition_view(i));
+
+    // Check if hypersparse segment is present in the partition
+    auto segment_offsets = graph_view.get_local_adj_matrix_partition_segment_offsets(i);
+    auto use_dcs =
+      segment_offsets
+        ? ((*segment_offsets).size() > (detail::num_sparse_segments_per_vertex_partition + 1))
+        : false;
+
+    if (use_dcs) {
+      auto major_hypersparse_first =
+        matrix_partition.get_major_first() +
+        (*segment_offsets)[detail::num_sparse_segments_per_vertex_partition];
+      // Calculate degrees in csr region
+      auto csr_begin = local_degrees.begin() + partial_offset;
+      auto csr_end   = local_degrees.begin() + partial_offset +
```

Comment:
If we compute local major degrees, this is no longer necessarily a CSR but can be either CSR or CSC. Maybe use sparse_begin() and sparse_end() instead?